lib/ts/recipe/multifactorauth/recipe.ts

```diff
@@ -0,0 +1,512 @@
+        let factorIds: string[] = [];
+        for (const func of this.getAllFactorsFromOtherRecipesFunc) {
+            const factorIdsRes = func(tenantConfig);
+            factorIds = factorIds.concat(factorIdsRes.factorIds);
+        }
+        return factorIds;
+    };
+
+    getAllAvailableFirstFactorIds = (tenantConfig: TenantConfig) => {
+        let factorIds: string[] = [];
+        for (const func of this.getAllFactorsFromOtherRecipesFunc) {
+            const factorIdsRes = func(tenantConfig);
+            factorIds = factorIds.concat(factorIdsRes.firstFactorIds);
+        }
+        return factorIds;
+    };
+
+    addGetFactorsSetupForUserFromOtherRecipes = (func: GetFactorsSetupForUserFromOtherRecipesFunc) => {
+        this.getFactorsSetupForUserFromOtherRecipesFuncs.push(func);
+    };
+
+    validateForMultifactorAuthBeforeFactorCompletion = async ({
+        tenantId,
+        factorIdInProgress,
+        session,
+        userLoggingIn,
+        isAlreadySetup,
+        signUpInfo,
+        userContext,
+    }: {
+        req: BaseRequest;
+        res: BaseResponse;
+        tenantId: string;
+        factorIdInProgress: string;
+        session?: SessionContainerInterface;
+        userLoggingIn?: User;
+        isAlreadySetup?: boolean;
+        signUpInfo?: {
+            email: string;
+            isVerifiedFactor: boolean;
+        };
+        userContext: UserContext;
+    }): Promise<{ status: "OK" } | MFAFlowErrors> => {
+        const tenantInfo = await Multitenancy.getTenant(tenantId, userContext);
+        const { status: _, ...tenantConfig } = tenantInfo!;
+        const validFirstFactors =
+            tenantInfo?.firstFactors || this.config.firstFactors || this.getAllAvailableFirstFactorIds(tenantConfig);
+
+        if (session === undefined) {
+            // No session exists, so we need to check if it's a valid first factor before proceeding
+            if (!validFirstFactors.includes(factorIdInProgress)) {
+                return {
+                    status: "DISALLOWED_FIRST_FACTOR_ERROR",
+                };
+            }
+            return {
+                status: "OK",
+            };
+        }
+
+        let sessionUser;
+        if (userLoggingIn) {
+            if (userLoggingIn.id !== session.getUserId()) {
+                // the user trying to login is not linked to the session user, based on session behaviour
+                // we just return OK and do nothing or replace replace the existing session with a new one
+                // we are doing this because we allow factor setup only when creating a new user
+
+                // this can happen when you got into login screen with an existing session and tried to log in with a different credentials
+                // or a case while doing secondary factor for phone otp but the user created a different account with the same phone number
+                return {
+                    status: "FACTOR_SETUP_NOT_ALLOWED_ERROR",
```

Comment:
It is only allowed during sign up you are right, but there is no guarantee the user isn't trying to set up an OTP factor using a phone/email address that already belongs to a pwless user.
This is in fact signifying the setup only during sign up limitation:
- This is an existing user signing in so we are not doing linking
- The user is not linked to the current user so we'd have to link it before marking as completed, which is not allowed.
- In fact the factor setup only by sign up is actually a limitation mostly because we do not want to do linking for existing users.

`otp-phone` and `otp-email` is considered to be set up if the user has the current email address as one of their addresses.

Reply:
we now use 403 in these cases